pkg/runtime/types.go

```diff
@@ -120,26 +132,370 @@ type Unknown struct {
+				strMap[k] = str
+			}
+		}
+		return strMap
+	}
+	return nil
+}
+
+func setNestedField(obj map[string]interface{}, value interface{}, fields ...string) {
+	m := obj
+	if len(fields) > 1 {
+		for _, field := range fields[0 : len(fields)-1] {
+			if _, ok := m[field].(map[string]interface{}); !ok {
+				m[field] = make(map[string]interface{})
+			}
+			m = m[field].(map[string]interface{})
+		}
+	}
+	m[fields[len(fields)-1]] = value
+}
+
+func setNestedMap(obj map[string]interface{}, value map[string]string, fields ...string) {
+	m := make(map[string]interface{}, len(value))
+	for k, v := range value {
+		m[k] = v
+	}
+	setNestedField(obj, m, fields...)
+}
+
+func (u *Unstructured) setNestedField(value interface{}, fields ...string) {
+	if u.Object == nil {
+		u.Object = make(map[string]interface{})
+	}
+	setNestedField(u.Object, value, fields...)
+}
+
+func (u *Unstructured) setNestedMap(value map[string]string, fields ...string) {
+	if u.Object == nil {
+		u.Object = make(map[string]interface{})
+	}
+	setNestedMap(u.Object, value, fields...)
+}
+
+// TODO: rewrite this.
+func setField(src reflect.Value, fieldName string, dest interface{}) error {
+	field := src.FieldByName(fieldName)
+	if !field.IsValid() {
+		return fmt.Errorf("%v does not have field %s", src, fieldName)
+	}
+	v, err := conversion.EnforcePtr(dest)
+	if err != nil {
+		return err
+	}
+	if field.Type().AssignableTo(v.Type()) {
+		v.Set(field)
+		return nil
+	}
+	if field.Type().ConvertibleTo(v.Type()) {
+		v.Set(field.Convert(v.Type()))
+		return nil
+	}
+	return fmt.Errorf("couldn't assign/convert %v to %v", field.Type(), v.Type())
+}
+
+func extractOwnerReference(src interface{}) (metatypes.OwnerReference, error) {
+	var ret metatypes.OwnerReference
+	v := src.(map[string]interface{})
+	ret.Kind = getNestedString(v, "kind")
+	ret.Name = getNestedString(v, "name")
+	ret.APIVersion = getNestedString(v, "apiVersion")
+	ret.UID = (types.UID)(getNestedString(v, "uid"))
+	return ret, nil
+}
+
+func setOwnerReference(dest interface{}, src metatypes.OwnerReference) error {
+	v, ok := dest.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("dest %v has to be of type map[string]interface()", dest)
+	}
+	setNestedField(v, src.Kind, "kind")
+	setNestedField(v, src.Name, "name")
+	setNestedField(v, src.APIVersion, "apiVersion")
+	setNestedField(v, string(src.UID), "uid")
+	return nil
+}
+
+func getOwnerReferences(object map[string]interface{}) (ownerReferences []map[string]interface{}, err error) {
+	field := getNestedField(object, "metadata", "ownerReferences")
+	if field == nil {
+		return ownerReferences, fmt.Errorf("cannot find field metadata.ownerReferences in %v", object)
+	}
+	ownerReferences, ok := field.([]map[string]interface{})
+	if !ok {
+		return ownerReferences, fmt.Errorf("expect metadata.ownerReferences to be a slice in %v", object)
+	}
+	return
+}
+
+func (u *Unstructured) GetOwnerReference(i int) (metatypes.OwnerReference, error) {
```

Comment:
I don't think this will be a huge performance hit. What's the expected number of owners? 1? 5? 20? I wouldn't worry about it at this point. I think it's a premature optimization that is leading to a fairly clunky API that is replacing language built-ins.

I wrote a benchmark, and I would say it's only a marginal difference: https://play.golang.org/p/mFMrwFh4Ix

output I got
```console
go run bench.go
     BenchAccessAllDirect(  1): 10000000	       142 ns/op
       BenchAccessAllCopy(  1):  5000000	       249 ns/op
--------------
     BenchAccessAllDirect(  2):  5000000	       264 ns/op
       BenchAccessAllCopy(  2):  3000000	       443 ns/op
--------------
     BenchAccessAllDirect(  5):  2000000	       641 ns/op
       BenchAccessAllCopy(  5):  1000000	      1027 ns/op
--------------
     BenchAccessAllDirect( 10):  1000000	      1285 ns/op
       BenchAccessAllCopy( 10):  1000000	      1889 ns/op
--------------
     BenchAccessAllDirect( 20):   500000	      2416 ns/op
       BenchAccessAllCopy( 20):   500000	      3696 ns/op
--------------
     BenchAccessAllDirect( 50):   200000	      6078 ns/op
       BenchAccessAllCopy( 50):   200000	      9000 ns/op
--------------
     BenchAccessAllDirect(100):   100000	     12062 ns/op
       BenchAccessAllCopy(100):   100000	     17742 ns/op
--------------
```